File: apps/web/src/lib/bridge/commands/governance.ts
/** Bridge commands — the project trust dashboard (issue #399): the repo-scoped,
 *  computed-on-demand governance summary and its shields-compatible badge export.
 *
 *  `workflow::trust` answers "this merge"; these answer "this repo, this month".
 *  Every value is aggregated per request from the task store, the flight-recorder
 *  ledgers and the append-only governance journal — nothing here is cached on
 *  either side of the bridge, so the dashboard can never drift from the journal.
 *  Outside the Tauri webview (browser preview / stories) they resolve deterministic
 *  mocks so the card renders instead of rejecting. */
import { invoke } from '@tauri-apps/api/core';

import { isTauri, tauriInvoke } from '../internal';
import type { ProjectTrustSummary } from '../types';

/** The browser-preview summary: a project with real history, so the card's stat
 *  grid, badge and journal feed all render in Storybook. Kept local (the shared
 *  `mocks.ts` is at its size cap), mirroring `./policy`'s activity mock. */
const MOCK_PROJECT_TRUST: ProjectTrustSummary = {
  generatedAt: '2026-07-29T14:20:00Z',
  merges: { tasks: 46, merged: 31, verified: 29, verifiedMerges: 28 },
  gauntlet: { runs: 34, passed: 32, passRate: 32 / 34 },
  guardrails: {
    toolsEvaluated: 4182,
    allowed: 4106,
    asked: 41,
    denied: 35,
    policyDenials: 22,
    sessions: 46,
    topRules: [
      { ruleId: 'harness-protected-path', count: 14, source: 'policy' },
      { ruleId: 'harness-bash-deny', count: 8, source: 'policy' },
      { ruleId: 'pipe-to-shell', count: 7, source: 'builtin' },
      { ruleId: 'workspace-confinement', count: 6, source: 'builtin' },
    ],
  },
  spend: { costUsd: 41.87, tasksWithCost: 39 },
  journal: {
    events: 12,
    quarantines: 2,
    policySaves: 5,
    arms: 3,
    disarms: 1,
    ratchets: 1,
    other: 0,
    corruptLines: 0,
    lastEventAt: '2026-07-29T13:58:11Z',
    recent: [
      {
        id: '11',
        ts: '2026-07-29T13:58:11Z',
        kind: 'quarantine',
        summary: 'quarantined 1 path(s) from agent reads',
        detail: ['docs/vendor/CHANGELOG.md'],
      },
      {
        id: '10',
        ts: '2026-07-29T13:57:40Z',
        kind: 'policy-save',
        summary:
          'policy saved — armed, 4 protected path(s), 3 bash denial(s), 2 denied read(s), tools 1/2/0 (deny/ask/allow)',
        detail: [],
      },
      {
        id: '9',
        ts: '2026-07-28T09:14:02Z',
        kind: 'arm',
        summary: 'armed check `structure-lock` (lint-meta)',
        detail: ['structure-lock'],
      },
      {
        id: '8',
        ts: '2026-07-27T16:02:55Z',
        kind: 'ratchet',
        summary: 'ratchet baseline snapshotted — 41 any, 5 @ts-ignore, 3 eslint-disable',
        detail: [],
      },
    ],
  },
  badge: {
    schemaVersion: 1,
    label: 'governance',
    message: '28 verified merges · 94% gauntlet · 35 denials',
    color: 'green',
  },
};

/** The active project's governance posture: verified merges, gauntlet pass rate,
 *  guardrail denials with rule attribution, spend, and the governance journal
 *  rolled up with a recent feed. Computed server-side per call. */
export async function projectTrustSummary(): Promise<ProjectTrustSummary> {
  return tauriInvoke<ProjectTrustSummary>('project_trust_summary', {}, MOCK_PROJECT_TRUST);
}

/** The shields.io endpoint payload for the active project as pretty JSON — the
 *  bytes a repo publishes so `https://img.shields.io/endpoint?url=…` renders its
 *  governance posture. Derived from the same summary above. */
export async function governanceBadgeJson(): Promise<string> {
  return tauriInvoke<string>(
    'governance_badge_json',
    {},
    JSON.stringify(MOCK_PROJECT_TRUST.badge, null, 2),
  );
}

/** The outcome of a badge export: whether the user completed the save, and where. */
export interface GovernanceBadgeExport {
  /** True once the JSON was written to `path`; false when the user cancelled the
   *  native save dialog or the call ran outside Tauri. */
  saved: boolean;
  /** The absolute path written to, or `null` when nothing was saved. */
  path: string | null;
}

/** Export the badge to a user-chosen `*.json` file: open the native save dialog,
 *  then have Rust re-compute and atomically write the payload (`write_governance_badge`).
 *  Keeping the write Rust-side preserves the ONE serializer while the path choice
 *  stays a native dialog; the backend refuses any destination inside a
 *  `.nightcore/` directory, so a badge can never land on the journal it reports on.
 *  Resolves `{ saved: false }` when the user cancels or outside Tauri. */
export async function exportGovernanceBadge(
  suggestedName: string = 'governance-badge',
): Promise<GovernanceBadgeExport> {
  if (!isTauri()) return { saved: false, path: null };
  // Dynamic import inside the isTauri() branch (the bridge idiom) — a static
  // `import { save }` at module scope breaks sibling tests whose
  // `vi.mock('@tauri-apps/plugin-dialog')` factory predates this export.
  const { save } = await import('@tauri-apps/plugin-dialog');
  const dest = await save({
    title: 'Export governance badge',
    defaultPath: `${suggestedName}.json`,
    filters: [{ name: 'JSON', extensions: ['json'] }],
  });
  if (dest === null) return { saved: false, path: null };
  await invoke<void>('write_governance_badge', { destPath: dest });
  return { saved: true, path: dest };
}

File: apps/desktop/src-tauri/src/commands/governance.rs
//! The project trust dashboard commands (issue #399) — thin shells over
//! `crate::workflow::project_trust`.
//!
//! Each resolves the ACTIVE project server-side (never a caller-supplied path, so
//! the webview cannot point the reader at an arbitrary repo — the `commands::policy`
//! posture), reads the task store, the flight-recorder ledgers and the governance
//! journal, and returns the COMPUTED summary. Nothing here persists the summary.
//!
//! The bodies read one file per task ledger, so they run on the blocking pool via
//! `spawn_blocking` — a synchronous `#[tauri::command]` would freeze the WKWebView
//! for the read (`reference_tauri_command_threading`). State is re-acquired with
//! `try_state` inside the 'static closure (a `State<'_>` guard cannot cross into
//! it), so an unmanaged store fails gracefully.

use std::path::Path;

use tauri::{AppHandle, Manager};

use crate::infra::path_confine::validate_export_dest;
use crate::store::TaskStore;
use crate::workflow::project_trust::{badge_of, build_summary, ProjectTrustSummary};

/// The active project's path via `try_state` (blocking-pool safe).
fn active_project_path(app: &AppHandle) -> Result<String, String> {
    let projects = app
        .try_state::<crate::project::ProjectStore>()
        .ok_or_else(|| "project store unavailable".to_string())?;
    projects
        .active()
        .map(|p| p.path)
        .ok_or_else(|| "no active project".to_string())
}

/// Resolve the inputs and compute the summary. Shared by both read commands so the
/// dashboard and the exported badge are computed from ONE call path.
fn summary_blocking(app: &AppHandle) -> Result<ProjectTrustSummary, String> {
    let root = active_project_path(app)?;
    let root = Path::new(&root);
    // An unmanaged task store degrades to "no tasks" rather than dropping the whole
    // dashboard — the guardrail + journal halves are still real evidence.
    let tasks: Vec<crate::task::Task> = app
        .try_state::<TaskStore>()
        .map(|store| store.list().iter().map(|t| (**t).clone()).collect())
        .unwrap_or_default();
    Ok(build_summary(
        &tasks,
        &crate::store::ledger::ledger_dir(root),
        root,
        crate::task::now_ms(),
    ))
}

/// The ACTIVE project's governance posture: verified merges, gauntlet pass rate,
/// guardrail denials, spend, and the governance journal rolled up.
#[tauri::command]
pub async fn project_trust_summary(app: AppHandle) -> Result<ProjectTrustSummary, String> {
    tauri::async_runtime::spawn_blocking(move || summary_blocking(&app))
        .await
        .map_err(|e| format!("project trust summary failed to run: {e}"))?
}

/// The shields.io endpoint payload for the ACTIVE project, as pretty JSON — the
/// bytes a repo publishes so `https://img.shields.io/endpoint?url=…` renders its
/// governance posture. Derived from the same summary the dashboard renders.
#[tauri::command]
pub async fn governance_badge_json(app: AppHandle) -> Result<String, String> {
    tauri::async_runtime::spawn_blocking(move || {
        let summary = summary_blocking(&app)?;
        badge_json(&summary)
    })
    .await
    .map_err(|e| format!("governance badge failed to run: {e}"))?
}

/// Write the badge JSON to a user-chosen path (the native save dialog supplies it).
///
/// `dest_path` is USER-CHOSEN and untrusted: it must be absolute and must not
/// descend through any `.nightcore/` directory
/// ([`validate_export_dest`]) — a badge written over
/// `.nightcore/ledger/project.ndjson` would destroy the journal it reports on. The
/// bytes are rendered here (one canonical serializer) and written atomically via
/// `store::write_atomic`, the `write_trust_report` idiom.
#[tauri::command]
pub async fn write_governance_badge(app: AppHandle, dest_path: String) -> Result<(), String> {
    tauri::async_runtime::spawn_blocking(move || {
        let dest = validate_export_dest(&dest_path)?;
        let summary = summary_blocking(&app)?;
        let json = badge_json(&summary)?;
        crate::store::write_atomic(&dest, json.as_bytes()).map_err(|e| {
            format!(
                "failed to write the governance badge to {}: {e}",
                dest.display()
            )
        })
    })
    .await
    .map_err(|e| format!("write governance badge failed to run: {e}"))?
}

/// Serialize the summary's badge as the shields endpoint payload. The ONE
/// serializer — the clipboard copy and the file export render identical bytes.
fn badge_json(summary: &ProjectTrustSummary) -> Result<String, String> {
    serde_json::to_string_pretty(badge_of(summary))
        .map_err(|e| format!("failed to serialize the governance badge: {e}"))
}

#[cfg(test)]
mod tests {
    use super::*;

    /// The published bytes are the shields ENDPOINT contract: `schemaVersion` (not
    /// `schema_version`), `label`, `message`, `color`. A rename here silently breaks
    /// every badge already pointed at a published file, so pin it.
    #[test]
    fn the_badge_serializes_as_a_shields_endpoint_payload() {
        let tmp = tempfile::TempDir::new().expect("temp dir");
        let root = tmp.path();
        let summary = crate::workflow::project_trust::build_summary(
            &[],
            &crate::store::ledger::ledger_dir(root),
            root,
            0,
        );

        let json = badge_json(&summary).expect("serialize");
        let value: serde_json::Value = serde_json::from_str(&json).expect("valid JSON");
        assert_eq!(value["schemaVersion"], 1);
        assert_eq!(value["label"], "governance");
        assert_eq!(value["message"], "not measured");
        assert_eq!(value["color"], "lightgrey");
        assert_eq!(
            value.as_object().expect("object").len(),
            4,
            "the endpoint payload carries exactly the four shields keys: {json}"
        );
    }
}
